onnxruntime/core/optimizer/bias_softmax_fusion.cc

```diff
@@ -0,0 +1,226 @@
+#include "core/optimizer/bias_softmax_fusion.h"
+#include "core/graph/graph_utils.h"
+#include "core/optimizer/utils.h"
+#include <deque>
+
+#define ILOGF(format_str, ...) \
+  LOGF_DEFAULT(INFO, format_str, ##__VA_ARGS__)
+
+using namespace ONNX_NAMESPACE;
+using namespace ::onnxruntime::common;
+namespace onnxruntime {
+
+// helper to check dimensions match on concrete or symbolic value
+bool operator!=(
+  const ONNX_NAMESPACE::TensorShapeProto_Dimension& lhs,
+  const ONNX_NAMESPACE::TensorShapeProto_Dimension& rhs) {
+    return !(lhs == rhs);
+} 
+
+bool operator==(const ONNX_NAMESPACE::TensorShapeProto_Dimension& lhs, int value) {
+  return utils::HasDimValue(lhs) && lhs.dim_value() == value;
+}
+
+bool operator!=(const ONNX_NAMESPACE::TensorShapeProto_Dimension& lhs, int value) {
+  return !(lhs == value);
+}
+
+void select_input_on_lhs_condition(bool lhs_condition, Node& add_node, NodeArg** input, NodeArg** mask) {
+  if (lhs_condition) {
+    *input = add_node.MutableInputDefs()[0];
+    *mask = add_node.MutableInputDefs()[1];
+  }
+  else {
+    *input = add_node.MutableInputDefs()[1];
+    *mask = add_node.MutableInputDefs()[0];
+  }
+}
+
+Status BiasSoftmaxFusion::ApplyImpl(Graph& graph, bool& modified, int graph_level, const logging::Logger& logger) const {
+  GraphViewer graph_viewer(graph);
+  const auto& node_topology_list = graph_viewer.GetNodesInTopologicalOrder();
+
+  // only support CUDA execution provider
+  auto& cep = GetCompatibleExecutionProviders();
+  if (cep.size() > 0 && cep.find(kCudaExecutionProvider) == cep.end())
+    return Status::OK();
+
+  for (auto node_index : node_topology_list) {
+    auto* node_ptr = graph.GetNode(node_index);
+    if (nullptr == node_ptr)
+      continue;  // node was removed
+
+    auto& node = *node_ptr;
+
+    ORT_RETURN_IF_ERROR(Recurse(node, modified, graph_level, logger));
+
+    // pattern match on dropout(softmax(input + mask)) subgraph
+    // -----------------------------------------------------------------------
+
+    // check node is add and has single output
+    if (!graph_utils::IsSupportedOptypeVersionAndDomain(node, "Add", {7}) ||
+        !graph_utils::IsSupportedProvider(node, {kCudaExecutionProvider}) ||
+        !optimizer_utils::CheckOutputEdges(graph, node, 1)) {
+      continue;
+    }
+
+    // check shape information is not available for both add inputs
+    auto& add_node = node;
+    NodeArg* input1 = add_node.MutableInputDefs()[0];
+    NodeArg* input2 = add_node.MutableInputDefs()[1];
+    const TensorShapeProto* S1 = input1->Shape();
+    const TensorShapeProto* S2 = input2->Shape();
+    if (S1 == nullptr || S2 == nullptr || S1->dim_size() < 1 || S2->dim_size() < 1) {
+      continue;
+    }
+
+    // check add is only consumed by softmax with matching exec provider
+    auto p = add_node.OutputNodesBegin();
+    if (p == add_node.OutputNodesEnd()) {
+      continue;
+    }
+    Node& softmax_node = const_cast<Node&>(*p);
```

Comment:
`const_cast` can be avoided with `graph.GetNode(p->Index())` which is admittedly also not that pretty